kvserver: prevent consistency queue from immediately running on new ranges

iskettaneh opened the pull request:
This commit stops the current behaviour of making new ranges immediately eligible for the consistency checker queue. This helps to improve the performance while ranges are being split.

It does this by setting the last consistency run timestamp to "now" instead of 0 initially. After that, this value is copied from the LHS to the RHS on splits, and the LHS remains on merged. This ensures two things:

1. We don't immediately make new ranges eligible for the consistency queue.

2. We don't indifferently ignore running the consistency queue on some part of the keyspace. This could have happened if we initialize the last queue timestamp to "now" for the RHS upon splits. Corner cases might happen that a range keeps splitting once before the 24 hours end, making part of the keyspace ineligible for consistency checks indifferently.

Sysbench before/after (almost the same exact performance as expected):

```
│ Metric        │ old.txt        │ new.txt        │ vs base         │
├───────────────┼────────────────┼────────────────┼──────────────-──┤
│ queries/sec   │ 30.00k ± 2%    │ 29.94k ± 1%    │ ~ (p=0.994 n=30)│
│ txns/sec      │ 1.500k ± 2%    │ 1.497k ± 1%    │ ~ (p=0.994 n=30)│
│ ms/min        │ 11.84 ± 2%     │ 11.84 ± 2%     │ ~ (p=0.959 n=30)│
│ ms/avg        │ 42.67 ± 2%     │ 42.75 ± 1%     │ ~ (p=0.994 n=30)│
│ ms/p95        │ 66.25 ± 5%     │ 66.84 ± 2%     │ ~ (p=0.586 n=30)│
│ ms/max        │ 263.3 ± 21%    │ 247.6 ± 18%    │ ~ (p=0.697 n=30)│
```

Fixes: #143136

Release note: None

iskettaneh committed:
kvserver: prevent consistency queue from immediately running on new ranges

This commit stops the current behaviour of making new ranges immediately
eligible for the consistency checker queue. This helps to improve the
performance while ranges are being split.

It does this by setting the last consistency run timestamp to "now"
instead of 0 initially. After that, this value is copied from the LHS to
the RHS on splits, and the LHS remains on merged. This ensures two
things:

1. We don't immediately make new ranges eligible for the consistency
queue.

2. We don't indifferently ignore running the consistency queue on some
part of the keyspace. This could have happened if we initialize the last
queue timestamp to "now" for the RHS upon splits. Corner cases might
happen that a range keeps splitting once before the 24 hours end, making
part of the keyspace ineligible for consistency checks indifferently.

Sysbench before/after (almost the same exact performance as expected):

```
│ Metric        │ old.txt        │ new.txt        │ vs base         │
├───────────────┼────────────────┼────────────────┼──────────────-──┤
│ queries/sec   │ 30.00k ± 2%    │ 29.94k ± 1%    │ ~ (p=0.994 n=30)│
│ txns/sec      │ 1.500k ± 2%    │ 1.497k ± 1%    │ ~ (p=0.994 n=30)│
│ ms/min        │ 11.84 ± 2%     │ 11.84 ± 2%     │ ~ (p=0.959 n=30)│
│ ms/avg        │ 42.67 ± 2%     │ 42.75 ± 1%     │ ~ (p=0.994 n=30)│
│ ms/p95        │ 66.25 ± 5%     │ 66.84 ± 2%     │ ~ (p=0.586 n=30)│
│ ms/max        │ 263.3 ± 21%    │ 247.6 ± 18%    │ ~ (p=0.697 n=30)│
```

Fixes: #143136

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145455)
<!-- Reviewable:end -->

iskettaneh commented:
I investigated the failure `TestLeaseTransfersUseExpirationLeasesAndPromoteCorrectly`. This test transfers a lease and waits for it to get promoted to a leader lease, while disabling `**AutomaticLeaseRenewal**`. The test for some reason now depends on the consistency checker to queue in and run the replica after the lease transfer happens. The consistency queue will end up promoting the lease to a leader lease [here](https://github.com/cockroachdb/cockroach/blob/308665f4fb5176f8f7dcefdbfe8b9eee669e9f5a/pkg/kv/kvserver/queue.go#L1061-L1065). 

@tbg The test probably needs to be updated. But you could ignore this failure during the review until I figure out what needs to be done with the test exactly.

tbg reviewed (commented):
Thank you for getting this out so early! Looks mostly good, I just hope you can "boil down" the test to be less expensive.

Re: the benchmarks, I think seeing that you've shown that the same benchmark showed a ~2% regression (though it ended up not being determined significant) and now it's far from showing any regression plus we know from the unit tests that the consistency checker just won't run for the first 24h, I think we're done here. Generally speaking though, since the roachtest benchmark uses a three minute warm-up interval, the benchmark alone might have not even captured the impact of the consistency checker, since the checks may have mostly finished in the warm-up interval. Anyway, no point in spending that extra work now, since we are already confident the consistency checks are no longer an issue.

iskettaneh committed:
kvserver: remove consistency checker dependency from lease transfer and promotion test

This commit removes the current unexpected dependency on the consistency
checker for leader leases in the test:
TestLeaseTransfersUseExpirationLeasesAndPromoteCorrectly.
Leader lease promotion after a lease transfer isn't immediate, and the
test is only passing because the lease gets promoted when the replica
runs on the consistency checker queue.

Epic: None

Release note: None

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145455)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->
